subsys/trusted_storage/src/auth_crypt_backend.c

```diff
@@ -0,0 +1,303 @@
+		if (status != PSA_SUCCESS) {
+			return -EIO;
+		}
+
+		status = get_random_cb((void *)&g_iv_high, sizeof(g_iv_high));
+		if (status != PSA_SUCCESS) {
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+psa_status_t trusted_storage_auth_crypt_backend_get(psa_storage_uid_t uid,
+				const char *prefix, size_t data_size,
+				size_t data_offset, size_t data_length,
+				void *p_data, size_t *p_data_length,
+				psa_storage_create_flags_t create_flags,
+				size_t (*get_encrypted_size_cb)(
+					size_t data_size),
+				psa_status_t (*crypt_cb)(
+					const void *key_buf, size_t key_len,
+					const void *iv_buf, size_t iv_len,
+					const void *add_buf, size_t add_len,
+					const void *input_buf, size_t input_len,
+					void *output_buf,
+					size_t output_size,
+					size_t *output_len))
+{
+	struct aead_additional_data additional_data;
+	uint8_t key_buf[AEAD_KEY_SIZE + 1];
+	uint8_t iv[AEAD_IV_SIZE];
+	size_t object_data_size;
+	size_t aead_out_size;
+	psa_status_t status;
+	int ret;
+
+	/* Calculate the exact output size of encrypted buffer */
+	object_data_size = get_encrypted_size_cb(data_size);
+
+	if (object_data_size > AEAD_MAX_BUF_SIZE) {
+		return PSA_ERROR_NOT_SUPPORTED;
+	}
+
+	ret = trusted_storage_get_object(uid, prefix,
+					 TRUSTED_STORAGE_FILENAME_SUFFIX_IV,
+					 iv, sizeof(iv));
+	if (ret == -ENOENT || ret == -ENODATA) {
+		return PSA_ERROR_DATA_CORRUPT;
+	} else if (ret < 0) {
+		return PSA_ERROR_STORAGE_FAILURE;
+	}
+
+	ret = trusted_storage_get_object(uid, prefix,
+					 TRUSTED_STORAGE_FILENAME_SUFFIX_DATA,
+					 aead_buf, object_data_size);
+	if (ret == -ENOENT || ret == -ENODATA) {
+		return PSA_ERROR_DATA_CORRUPT;
+	} else if (ret < 0) {
+		return PSA_ERROR_STORAGE_FAILURE;
+	}
+
+	/* Key is ASCII version of UID */
+	snprintf(key_buf, AEAD_KEY_SIZE + 1, AEAD_KEY_PATTERN,
+		 (unsigned int)((uid) >> 32),
+		 (unsigned int)((uid) & 0xffffffff));
+
+	additional_data.uid = uid;
+	additional_data.flags = create_flags;
+	additional_data.size = data_length;
+
+	status = crypt_cb(key_buf, AEAD_KEY_SIZE, iv, AEAD_IV_SIZE,
+			 (void *)&additional_data, sizeof(additional_data),
+			 aead_buf, object_data_size,
+			 data_buf, TRUSTED_STORAGE_MAX_ASSET_SIZE,
+			 &aead_out_size);
+
+	memset(key_buf, 0, AEAD_KEY_SIZE);
+	memset(iv, 0, sizeof(iv));
+	memset(&additional_data, 0, sizeof(additional_data));
+	memset(aead_buf, 0, AEAD_MAX_BUF_SIZE);
+
+	if (status != PSA_SUCCESS) {
+		return status;
+	}
+
+	if ((data_offset + data_length) > aead_out_size) {
+		memset(data_buf, 0, TRUSTED_STORAGE_MAX_ASSET_SIZE);
+		return PSA_ERROR_INVALID_SIGNATURE;
+	}
+
+	memcpy(p_data, data_buf + data_offset, data_length);
+
+	memset(data_buf, 0, TRUSTED_STORAGE_MAX_ASSET_SIZE);
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t trusted_storage_auth_crypt_backend_set(psa_storage_uid_t uid,
+				const char *prefix, size_t data_length,
+				const void *p_data,
+				psa_storage_create_flags_t create_flags,
+				psa_status_t (*crypt_cb)(
```

Comment:
I know the Zephyr coding standard is negative to typedefs, but I think that in this case the readability unfortunately suffers a lot. There are examples of function pointers for the type of callback you write here. Please change it to a typedef!